sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -133,12 +133,48 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
 
   private def visitCompoundBodyImpl(
       ctx: CompoundBodyContext,
-      label: Option[String]): CompoundBody = {
+      label: Option[String],
+      allowVarDeclare: Boolean): CompoundBody = {
     val buff = ListBuffer[CompoundPlanStatement]()
     ctx.compoundStatements.forEach(compoundStatement => {
       buff += visit(compoundStatement).asInstanceOf[CompoundPlanStatement]
     })
 
+    val compoundStatements = buff.toList
+
+    if (allowVarDeclare) {
+      val declareVarStatement = compoundStatements
+        .dropWhile(statement => statement.isInstanceOf[SingleStatement] &&
+          statement.asInstanceOf[SingleStatement].parsedPlan.isInstanceOf[CreateVariable])
+        .filter(_.isInstanceOf[SingleStatement])
+        .find(_.asInstanceOf[SingleStatement].parsedPlan.isInstanceOf[CreateVariable])
+
+      declareVarStatement match {
+        case Some(SingleStatement(parsedPlan)) =>
+          throw SqlScriptingErrors.variableDeclarationOnlyAtBeginning(
+            parsedPlan.asInstanceOf[CreateVariable]
+              .name.asInstanceOf[UnresolvedIdentifier]
+              .nameParts.last,
```

Comment:
we should use the full variable name, not just the last name part. We can use `DataTypeErrorsBase#toSQLId` to generate the name string

Reply:
We did it like this since we are still figuring out the best way to implement local variables, and thought of addressing this "issue" then. But, what you suggested seems like the thing that we will end up doing eventually, so no harm to do it now as well.

Reply:
Done